src/main/java/school/hei/haapi/service/UserService.java

```diff
@@ -37,24 +46,51 @@ public User getByEmail(String email) {
   public List<User> saveAll(List<User> users) {
     List<User> savedUsers = userRepository.saveAll(users);
     eventProducer.accept(users.stream()
-              .map(this::toTypedEvent)
-              .collect(toUnmodifiableList()));
+            .map(this::toTypedEvent)
+            .collect(toUnmodifiableList()));
     return savedUsers;
   }
 
   private TypedUserUpserted toTypedEvent(User user) {
     return new TypedUserUpserted(
-        new UserUpserted()
-            .userId(user.getId())
-            .email(user.getEmail())
+            new UserUpserted()
+                    .userId(user.getId())
+                    .email(user.getEmail())
     );
   }
 
   public List<User> getByRole(User.Role role, PageFromOne page, BoundedPageSize pageSize) {
     Pageable pageable = PageRequest.of(
-        page.getValue() - 1,
-        pageSize.getValue(),
-        Sort.by(ASC, "ref"));
+            page.getValue() - 1,
+            pageSize.getValue(),
+            Sort.by(ASC, "ref"));
     return userRepository.getByRole(role, pageable);
   }
+
+  public List<User> getByCriteria(User.Role role, PageFromOne page, BoundedPageSize pageSize,
+                                  String ref, String firstName, String lastName) {
+    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
```

Comment:
No. Create UserRepository::findByReferenceAndFirstNameAndLastName(ref, first, last).
Equality is `Like`. Default is `like %` for everything.